[SYCL2020] Group functions

DieGoldeneEnte opened the pull request:
This PR adds a naive implementation for group_functions to hipSYCL. This includes implementations of
* group_broadcast
* group_barrier
* group_{any,all,none]_of
* group_reduce
* group_{ex,in}clusive_scan

on CPU and NVIDIA/AMD GPUs, as well as tests for these functions. I will provide optimized versions in later PRs.
It also includes (group_)functions using two pointers (beginning/end), but these are not in the specification and are not meant to be used yet (as such they reside in the `detail` namespace).

At the moment all tests pass (except some problems with the pointer-based functions on CPU which sometimes fail, I am still investigating). [(private results)](https://github.com/hipSYCL/hipSYCL-ci-bot/actions/runs/387445691)

I would love to get some feedback. If you find some template-parameters or formatting you don't like, there is a chance I missed them in one of my cleanup/refactoring attempts. Just tell me so I can fix it. Some small changes like splitting the tests into multiple files for faster compilation might be added here,

DieGoldeneEnte committed:
Add initial group-functions

This adds the non-optimized group-functions and tests for
any_of/all_of/none_of, reduce, inclusive/exclusive scan for groups and
subgroups on CPU and AMD/NVIDIA GPUs.
It also contains similar functions which take a begin and end pointer.

Optimizations will follow in later commits.

illuhad commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,231 @@
+
+#ifndef HIPSYCL_LIBKERNEL_CUDA_GROUP_FUNCTIONS_HPP
+#define HIPSYCL_LIBKERNEL_CUDA_GROUP_FUNCTIONS_HPP
+
```
Should include `sub_group` header?

illuhad commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,231 @@
+
+template <typename T, int N>
+HIPSYCL_KERNEL_TARGET
+sycl::vec<T, N> group_broadcast(sub_group g, sycl::vec<T, N> x, typename sub_group::linear_id_type local_linear_id = 0) {
```
Line break? Maybe use something like clang-format

illuhad commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,231 @@
+
+template <typename T, typename std::enable_if<std::is_arithmetic<T>::value, T>::type * = nullptr>
+__device__
+T _shfl(T x, int id) {
```
We don't have a convention for `_identifier`. There are some `__identifier()` for compiler-glue stuff or low-level macros such as `__hipsycl_lid_x` etc. How about just using `shufl_impl`?

DieGoldeneEnte replied:
Do you really want `shufl_impl` or instead of `shfl_impl`/`shuffle_impl`? :)

illuhad replied:
:laughing: You're absolutely right, let's use `shuffle_impl`.

illuhad commented on include/hipSYCL/sycl/libkernel/hip/group_functions.hpp:
```diff
@@ -0,0 +1,246 @@
+
+#ifndef HIPSYCL_LIBKERNEL_HIP_GROUP_FUNCTIONS_HPP
+#define HIPSYCL_LIBKERNEL_HIP_GROUP_FUNCTIONS_HPP
+
```
Include `sub_group` header?

illuhad commented on include/hipSYCL/sycl/libkernel/group_functions.hpp:
```diff
@@ -0,0 +1,98 @@
+HIPSYCL_KERNEL_TARGET
+T group_reduce(Group g, V x, T init, BinaryOperation binary_op) {
+  T reduction = group_reduce(g, T{x}, binary_op);
+  return binary_op(reduction, init);;
```
double semicolon ;)

illuhad commented on include/hipSYCL/sycl/libkernel/host/group_functions.hpp:
```diff
@@ -0,0 +1,514 @@
+
+  group_barrier(g);
+
+  detail::free_shared_ptr(g, scratch);
```
We are in SYCL. We call it local memory, except when referring explicitly to HIP/CUDA and their `__shared__` memory :-)

illuhad commented on include/hipSYCL/sycl/libkernel/group_functions.hpp:
```diff
@@ -0,0 +1,98 @@
+#include "generic/hiplike/group_functions.hpp"
+#include "cuda/group_functions.hpp"
+#include "hip/group_functions.hpp"
+#include "host/group_functions.hpp"
```
Better to have includes at the beginning of a file. Is there a hard reason why you need them at the end?
Maybe it would also be clearer if you `#ifdef backend` the include here. You already `ifdef` inside the headers, but doing it here in addition might make it clearer what happens?

DieGoldeneEnte replied:
It was needed in an earlier version, I'll move it to the top and add the `#ifdef`.

illuhad commented on include/hipSYCL/sycl/libkernel/nd_item.hpp:
```diff
@@ -372,6 +374,7 @@ struct nd_item
   const range<dimensions> _local_range;
   const range<dimensions> _num_groups;
   const id<dimensions> _global_id;
+  void *_shared_ptr;
```
`_local_ptr` ?

illuhad commented on tests/sycl/group_functions.cpp:
```diff
@@ -0,0 +1,1800 @@
+using namespace cl;
+
+#ifdef TESTS_GROUPFUNCTION_FULL
+using test_types = boost::mpl::list<char, int, unsigned int,long long, float, double, sycl::vec<int, 1>, sycl::vec<int, 2>, sycl::vec<int, 3>, sycl::vec<int, 4>, sycl::vec<int, 8>,  sycl::vec<short, 16>, sycl::vec<long, 3>, sycl::vec<unsigned int, 3>>;
```
Line breaks, e.g. use clang-format

illuhad commented on tests/sycl/group_functions.cpp:
```diff
@@ -0,0 +1,1800 @@
+template<typename T, int N>
+std::string T2str(sycl::vec<T, N> v)
+{
+    std::stringstream ss{};
```
We use 2 spaces as indentation (`clang-format` can do it automatically for you  :-) )

illuhad commented on tests/sycl/group_functions.cpp:
```diff
@@ -0,0 +1,1800 @@
+template<int N, int M, typename T>
+class test_kernel;
+
+template<int TEST_ID, typename T, typename DataGenerator, typename FunctionUnderTest, typename TestFunction>
```
Why is `TEST_ID` capitalized?

DieGoldeneEnte replied:
missed it when renaming the other parameters, also changed it to `CallingLine`, since that's what I'm using as 'id' anyway.

illuhad commented on tests/sycl/group_functions.cpp:
```diff
@@ -0,0 +1,1800 @@
+            detail::check_binary_reduce<T, __LINE__>(vIn, local_size, global_size, std::vector<bool>{false, false, false, true}, "all_of");
+        };
+
+        test_nd_group_function_1d<__LINE__, T>(local_size, global_size,
```
It's an interesting idea to use `__LINE__` as kernel name. I'm torn between thinking that it's genius or evil C-macro hacking ;)

DieGoldeneEnte replied:
Since I call the kernel in the function I need some way to get unique names, the type alone is not enough. If you have another idea I'll glady use it. Once we use c++20 this can be replaced with [std::source_location::current()](https://en.cppreference.com/w/cpp/utility/source_location/current).

illuhad replied:
The traditional way is to just name it: `test_nd_group_function_1d<class my_test_case_name_or_a_description_of_it, T>(...)`. But I see that this can be cumbersome when there are many test cases.

DieGoldeneEnte replied:
I use line to make the tests as copy&past ready as possible. As an added bonus I get the line number if something goes wrong and I have to read mangled names :>

illuhad reviewed (commented):
Some initial feedback :)

illuhad commented:
Let's try @fodinabor 's awesome new feature:

@hipsycl-ci-bot test this please!

DieGoldeneEnte commented:
Do you know how I can stop clang-format from putting `HIPSYCL_KERNEL_TARGET` on the same line as the function?
I used it before, but since I need to reformat it by hand I only used it form the initial refactor :)

illuhad commented:
Let's try @fodinabor 's awesome new feature, take two:

@hipsycl-ci-bot test this again please!

EDIT: Retrigger!

DieGoldeneEnte commented:
Let's see if I am mighty enough for this (and manage to guess all the necessary conditions)

@hipsycl-ci-bot test this again please!

DieGoldeneEnte commented:
I managed to create a clang-format config, that somewhat resembles existing code. If you want to get HIPSYCL_KERNEL_TARGET, \_\_device\_\_, ... on it's own line you can use:
```
StatementMacros:
  - HIPSYCL_KERNEL_TARGET
  - HIPSYCL_UNIVERSAL_TARGET
  - __device__
```

illuhad commented:
Sorry, but since your PR is already the test bed, you will have to endure some spam ;)

@hipsycl-ci-bot test with actual GPUs, please :)

EDIT: Attempt 3

DieGoldeneEnte commented:
Added a barrier tests and fixed the race-condition (which was in the tests...). No the groups don't override the input of the others. I also split the tests leading to a reduction of ~1min on my machine.

@hipsycl-ci-bot test it with CUDA for me <3

fodinabor commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,240 @@
+namespace detail {
+
+template<typename T,
+         typename std::enable_if<std::is_arithmetic<T>::value, T>::type * = nullptr>
```
this is very C++11 style... `std::enable_if_t<std::is_arithmetic_v<T>, int> = 0>` could be used with C++17 :)

fodinabor commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,240 @@
+         typename std::enable_if<std::is_arithmetic<T>::value, T>::type * = nullptr>
+__device__
+T shuffle_impl(T x, int id) {
+  return __shfl_sync(0xFFFFFFFF, x, id);
```
maybe use a `constexpr std::uint32_t AllMask` or similar for the mask?

DieGoldeneEnte replied:
good Idea!

fodinabor commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,240 @@
+// reduce
+template<typename T, typename BinaryOperation>
+HIPSYCL_KERNEL_TARGET
+T group_reduce(sub_group g, T x, BinaryOperation binary_op) {
```
couldn't these _higher_ level implementations be shared between at least hiplike backends? looks like quite some code duplication .. or do you plan to possibly use some more intrinsics (like cuda's very new `__reduce_add_sync`)

DieGoldeneEnte replied:
I could try to merge them into hiplike, I'll take a look. I originally split them up, because the shuffle-instruction is different. Now we have `shuffle_impl`, so I could give this a try.

DieGoldeneEnte replied:
I'll keep them separated for now, since it is likely I will separate them again when optimizing. I'll keep it in mind though in case the optimized code can be implemented in a hiplike fashion :)

fodinabor commented on include/hipSYCL/sycl/libkernel/generic/hiplike/group_functions.hpp:
```diff
@@ -0,0 +1,591 @@
+
+// reduce implementation
+template<typename Group, typename T, typename BinaryOperation,
+         typename std::enable_if<!std::is_same<Group, sub_group>::value, Group>::type * = nullptr>
```
as above with `std::is_same_v` :)

fodinabor commented on include/hipSYCL/sycl/libkernel/cuda/group_functions.hpp:
```diff
@@ -0,0 +1,240 @@
+
+  auto return_value = detail::shuffle_impl(local_x, lid - 1);
+
+  if (lid == 0)
```
just from looking at the spec, using `g.leader()` might be beautiful here.. but shouldn't change any semantics and in theory, should be well optimizable.. so guess it just comes down to whether one likes the common pattern or the readable one

DieGoldeneEnte replied:
I'll have a look again. I know it isn't yet implemented for groups (I could/should just do it if necessary), so I opted for `lid==0`.

fodinabor commented on include/hipSYCL/sycl/libkernel/generic/hiplike/group_functions.hpp:
```diff
@@ -0,0 +1,591 @@
+    end_prt = last;
+
+  if (end_prt > start_ptr) {
+    auto local = binary_op(*start_ptr, *(start_ptr + 1));
```
what if `end_ptr == start_ptr + 1`? wouldn't `*(start_ptr + 1)` be an invalid access as the end_ptr is not in the range?
also typo in `end_prt` :)

DieGoldeneEnte replied:
correct I didn't think of this case in my last rewrite :/

fodinabor commented on include/hipSYCL/sycl/libkernel/hip/group_functions.hpp:
```diff
@@ -0,0 +1,255 @@
+  if constexpr (std::is_signed<T>::value) {
+    return __shfl(x, id);
+  } else {
+    return static_cast<T>(__shfl(static_cast<typename std::make_signed<T>::type>(x), id));
```
`static_cast<std::make_signed_t<T>>`

fodinabor commented on include/hipSYCL/sycl/libkernel/hip/group_functions.hpp:
```diff
@@ -0,0 +1,255 @@
+         typename std::enable_if<std::is_arithmetic<T>::value, T>::type * = nullptr>
+__device__
+T shuffle_impl(T x, int id) {
+  if constexpr (std::is_signed<T>::value) {
```
`is_signed_v`

fodinabor reviewed (commented):
Love it! :)
Mostly some readability notes that I just mentioned once and could be applied generally and one small uncertainty about cerrectness :)

DieGoldeneEnte commented:
@fodinabor 
Using the c++17 template-magic is a good point, didn't think about them at all ':)

DieGoldeneEnte committed:
Added leader to group

DieGoldeneEnte commented:
@illuhad the first of the new commits adds leader() to sycl::group. I could put it in a separate PR, but then I have to rebase this PR to use it. I think it should be OK to hide it in here. In case you want to squash this, I can squash manually and keep that PR separate, as well as make sure the commit message is useful (it will likely be a long list of one-liners that don't make much sense when squashing automatically).

DieGoldeneEnte commented:
@hipsycl-ci-bot test please

Edit: Once more please

fodinabor commented on tests/sycl/group_functions/group_functions.hpp:
```diff
@@ -53,7 +53,7 @@ using test_types = boost::mpl::list<char, unsigned int, float, double, sycl::vec
 
 template<typename T>
-using elementType = typename std::remove_reference<decltype(T{}.s0())>::type;
+using elementType = typename std::remove_reference_t<decltype(T{}.s0())>;
```
no need for `typename` before `_t` variants, it is only required for dependent names aka if you have a `::` after a type name (ignore those after namespace names... :)), see: https://en.cppreference.com/w/cpp/keyword/typename

DieGoldeneEnte replied:
good catch :+1:

illuhad commented:
@DieGoldeneEnte that's fine. If you want to squash manually feel free, otherwise I would just squash automatically.

DieGoldeneEnte commented:
@illuhad Ok, I'll do that, just tell me when you want to merge it. It might make sense if the group functions are not merged under the title "removed redundant typename" or similar :D

illuhad commented:
please squash to prepare for merging :)

DieGoldeneEnte commented:
Done :)

Only lost a newline while moving the group-commit to the front.
```
diff --git a/include/hipSYCL/sycl/libkernel/group.hpp b/include/hipSYCL/sycl/libkernel/group.hpp
index c26e1d8..b553522 100644
--- a/include/hipSYCL/sycl/libkernel/group.hpp
+++ b/include/hipSYCL/sycl/libkernel/group.hpp
@@ -478,7 +478,6 @@ private:
   const host_barrier_type* _group_barrier;
   const id_type _local_id;
   void *_local_memory_ptr;
-
 public:
 #endif
```

DieGoldeneEnte commented:
@hipsycl-ci-bot test please, llvm doesn't want to be installed on the default actions

Edit: again please

DieGoldeneEnte commented:
there seems to be a problem with CUDA-broadcasts, it happens in less than every 100th run...
I'll take a look what the cause could be.

Edit: I could only get the test to fail 2 out of 500 runs...

illuhad commented:
If you are using host accessors or rely on buffer writebacks for validation in your tests, then this might also be cause by a bugged task graph if you don't have PR #395 merged yet.

DieGoldeneEnte commented:
That might be the reason, I could only reproduce it once. I think the problem occurred under heavy CPU-load, but that is just a suspicion.

DieGoldeneEnte commented:
@hipsycl-ci-bot test this :snake: 

Edit: again after rebase

DieGoldeneEnte commented:
FunctionUnderTest ->  TestedFunction
TestFunction -> ValidationFunction
TCompare -> compare_type
TInit -> initialize_type
T2str -> type_to_string
and a bunch of camel-case to snake-case conversion :)

Edit: removed emojis, it looked too much like some PR for a js-library...

DieGoldeneEnte commented:
@illuhad I think this is ready for merging now, pretty sure I got everything.

illuhad commented:
In it goes, thank you so much for this awesome contribution :)
